src/System.Text.Primitives/System/Text/InvariantParser_byte.cs

```diff
@@ -29,42 +30,105 @@ public static bool TryParse(byte[] utf8Text, int index, out byte value, out int
             value = 0;
             bytesConsumed = 0;
 
-            for (int byteIndex = index; byteIndex < utf8Text.Length; byteIndex++) // loop through the byte array
+            if (cultureAndEncodingInfo.IsInvariantUtf8)
             {
-                byte nextByteVal = (byte)(utf8Text[byteIndex] - '0');
-                if (nextByteVal > 9) // if the next character is not a digit
+                for (int byteIndex = index; byteIndex < utf8Text.Length; byteIndex++) // loop through the byte array
                 {
-                    if (bytesConsumed == 0) // check to see if we've processed any digits at all
+                    byte nextByteVal = (byte)(utf8Text[byteIndex] - '0');
+                    if (nextByteVal > 9) // if the next character is not a digit
                     {
-                        value = default(byte); // if we haven't, set value to 0 and return false
+                        if (bytesConsumed == 0) // check to see if we've processed any digits at all
+                        {
+                            value = default(byte); // if we haven't, set value to 0 and return false
+                            return false;
+                        }
+                        else
+                        {
+                            return true; // otherwise return true
+                        }
+                    }
+                    else if (value > Byte.MaxValue / 10)
+                    {
+                        value = 0;
+                        bytesConsumed = 0;
                         return false;
                     }
-                    else
+                    else if (value > 0 && Byte.MaxValue - value * 10 < nextByteVal) // overflow
                     {
-                        return true; // otherwise return true
+                        value = 0;
+                        bytesConsumed = 0;
+                        return false;
                     }
+
+                    byte candidate = (byte)(value * 10 + nextByteVal); // left shift the value and add the nextByte.
+                    value = candidate;
+                    bytesConsumed++;
                 }
-                else if (value > Byte.MaxValue / 10)
-                {
-                    value = 0;
-                    bytesConsumed = 0;
-                    return false;
-                }
-                else if (value > 0 && Byte.MaxValue - value * 10 < nextByteVal) // overflow
+            }
+            else if (cultureAndEncodingInfo.IsInvariantUtf16)
```

Comment:
I was going to make the same comment. Unless I'm misunderstanding something, we should be able to loop through 16-bit chunks instead of looping over individual bytes, and it would probably eliminate some of this confusing logic.